File: julia-flat-matrix/TSPBenchmark/src/TSPBenchmark.jl
module TSPBenchmark

using CPUTime
using StaticArrays

export benchmark_many

struct TSPData{ T<:Real }
    n::Int    # number of nodes
    d::Vector{T} # distance matrix
end

mutable struct TSPSolution
    tour::Vector{Int}
end

function dist(d::TSPData{T}, i::T2, j::T2) where T where T2
    d.d[i * d.n - d.n + j]
end

function read_data(fname::String, T::DataType=Int)
    lines = open(fname) do f
        collect(eachline(f))
    end
    n = 0
    drow = 1
    d = Vector{T}()
    nsols = 0
    solutions = TSPSolution[]
    for line in lines
        line = replace(line, r"#.*" => "")
        tokens = split(line)
        if length(tokens) == 0
            continue
        elseif length(tokens) == 2 && n == 0
            n = parse(Int, tokens[1])
            nsols = parse(Int, tokens[2])
            d = zeros(T, n^2)
        elseif drow ≤ n
            row = [parse(T, x) for x in tokens]
            f = 1 + (n * (drow - 1))
            d[f:f + n - 1] = row
            drow += 1
        elseif length(solutions) ≤ nsols
            push!(solutions, TSPSolution([1+parse(Int, x) for x in tokens]))
        end
    end
    data = TSPData{T}(n, [x for x in d])
    return data, solutions
end

function two_opt(data::TSPData{T}, sol::TSPSolution) where T
    t = 0
    while first2eimprovement(data, sol)
        t += 1
    end
    return t
end

function first2eimprovement(d::TSPData{T}, sol::TSPSolution) where T
    for p1 in 1:length(sol.tour)-3, p2 in p1+2:length(sol.tour)-1
        if dist(d, sol.tour[p1], sol.tour[p1+1]) +
            dist(d, sol.tour[p2], sol.tour[p2+1]) >
            dist(d, sol.tour[p1], sol.tour[p2]) +
            dist(d, sol.tour[p1+1], sol.tour[p2+1])
            # improving 2-exchange found
            for i in 0:(p2-p1+1) ÷ 2 - 1
                sol.tour[p1+1+i], sol.tour[p2-i] =
                    sol.tour[p2-i], sol.tour[p1+1+i]
            end
            return true
        end
    end
    return false
end

function or_opt(data::TSPData{T}, sol::TSPSolution) where T
    t = 0
    while firstorimprovement(data, sol)
        t += 1
    end
    return t
end

function firstorimprovement(d::TSPData{T}, s::TSPSolution) where T
    for i in 2:length(s.tour)-1, l in 1:(min(3, length(s.tour)-i)),
        p in [ 1:i-2 ; i+l:length(s.tour)-1 ]
        #
        δ = dist(d, s.tour[i-1], s.tour[i+l]) + dist(d, s.tour[p], s.tour[i]) +
            dist(d, s.tour[i+l-1], s.tour[p+1]) -
            dist(d, s.tour[p], s.tour[p+1]) - dist(d, s.tour[i-1], s.tour[i]) -
            dist(d, s.tour[i+l-1], s.tour[i+l])
        # improvement found
        if δ < 0
            seq = [ k for k in s.tour[i:i+l-1] ]
            if i < p
                for j in i:p-l
                    s.tour[j] = s.tour[j+l]
                end
                for j in 1:l
                    s.tour[p+j-l] = seq[j]
                end
            else
                for j in i-1:-1:p+1
                    s.tour[j+l] = s.tour[j]
                end
                for j in 1:l
                    s.tour[p+j] = seq[j]
                end
            end
            return true
        end
    end
    return false
end

# old less efficient version

# function firstorimprovement(d::TSPData{T}, s::TSPSolution) where T
#     for i in 2:length(s.tour)-1, l in 1:(min(3, length(s.tour)-i)),
#         p in [ 1:i-2 ; i+l:length(s.tour)-1 ]
#         #
#         δ = d.d[s.tour[i-1], s.tour[i+l]] + d.d[s.tour[p], s.tour[i]] +
#             d.d[s.tour[i+l-1], s.tour[p+1]] -
#             d.d[s.tour[p], s.tour[p+1]] - d.d[s.tour[i-1], s.tour[i]] -
#             d.d[s.tour[i+l-1], s.tour[i+l]]
#         # improvement found
#         if δ < 0
#             if i < p
#                 s.tour = [ s.tour[1:i-1] ; s.tour[i+l:p] ; s.tour[i:i+l-1] ;
#                            s.tour[p+1:end] ]
#             else
#                 s.tour = [ s.tour[1:p] ; s.tour[i:i+l-1] ; s.tour[p+1:i-1] ;
#                            s.tour[i+l:end] ]
#             end
#             return true
#         end
#     end
#     return false
# end

function benchmark_one(data::TSPData{T}, solutions::Vector{TSPSolution},
                       benchmarkname::String) where T
    totaltime = 0.0
    nimpr = 0
    for s in solutions
        if benchmarkname == "2-opt"
            t = @CPUelapsed n = two_opt(data, s)
        elseif lowercase(benchmarkname) == "or-opt"
            t = @CPUelapsed n = or_opt(data, s)
        else
            throw(ArgumentError("Unknown benchmark: $benchmarkname"))
        end
        totaltime += t
        nimpr += n
    end
    return nimpr, totaltime
end

function benchmark_many(dirname::String, benchmarkname::String="2-opt")
#    println("#language,version,benchmark,instance,n,nsolutions,n_improvements,time(s)")
    nimpr = 0
    for fname in readdir(dirname)
        d, sols = read_data(joinpath(dirname, fname), Int)
        n, l = benchmark_one(d, sols, benchmarkname)
        println(join((basename(pwd()), VERSION, benchmarkname, fname,
                      string(d.n),
                      string(size(sols,1)), string(n), string(l)),
                     ","))
        nimpr += n
    end
end

end # module

File: implementations/javascript-nested-matrix/tspbenchmark.js
#!/usr/bin/env node

// class: TSP instance
function TSPData(n, d) {
    this.n = n;
    this.d = d;
}

// class: TSP Solution
function TSPSolution(data, sequence) {
    this.data = data;
    this.nodes = sequence;

    // perform first-improvement 2-opt
    this.twoOpt = function() {
	var t = 0;
	while(this.first2eImprovement()) {
	    t += 1;
	}
	return t;
    }

    // perform the first improving two-exchange found
    var tour = this.nodes;
    var d = this.data.d;
    this.first2eImprovement = function() {
	for(var p1=0; p1 < tour.length - 3; p1++) {
	    for(var p2=p1+2; p2 < tour.length - 1; p2++) {
		if (d[tour[p1]][tour[p1+1]] + d[tour[p2]][tour[p2+1]] >
		    d[tour[p1]][tour[p2]] + d[tour[p1+1]][tour[p2+1]]) {
		    // improving 2-exchange found
		    for(var i=0; i < Math.floor((p2-p1+1) / 2); i++) {
			[tour[p1+1+i], tour[p2-i]] = [tour[p2-i], tour[p1+1+i]];
		    }
		    return true;
		}
	    }
	}
	return false;
    };    

    // perform first-improvement 2-opt
    this.orOpt = function() {
	var t = 0;
	while(this.firstOrImprovement()) {
	    t += 1;
	}
	return t;
    }

    this.orDelta = function(i, l, p) {
	return d[tour[i-1]][tour[i+l]] + d[tour[p]][tour[i]]
	    + d[tour[i+l-1]][tour[p+1]] - d[tour[p]][tour[p+1]]
	    - d[tour[i-1]][tour[i]] - d[tour[i+l-1]][tour[i+l]];
    }
	
    // perform the first improving Or move found
    this.firstOrImprovement = function() {
	for(var i=1; i < tour.length - 1; i++) {
	    for(var l=1; l < 1 + Math.min(3, tour.length - 1 - i); l++) {
		for(var p=0; p < i-1; p++) {
		    if (this.orDelta(i,l,p) < 0) {
			let t = tour.splice(i, l);
			tour.splice(p+1, 0, ...t);
			return true;
		    }
		}
		for(var p=i+l; p < tour.length - 1; p++) {
		    if (this.orDelta(i,l,p) < 0) {
			let t = tour.splice(i, l);
			tour.splice(p+1-l, 0, ...t);
			return true;
		    }
		}
	    }
	}
	return false;
    };

}

// this version requires node
function fileAsLines(fname) {
    var fs = require('fs');
    return fs.readFileSync(fname).toString().split("\n");
}

function readData(fname) {
    var n = 0;
    var d = [];
    var tours = [];
    var nsols = 0;
    fileAsLines(fname).forEach(line => {
	var tokens = line.replace(/#.*/i, '').split(' ');
	if (tokens.length == 0) {
	    ;
	} else if (n == 0 && tokens.length == 2) {
	    n = parseInt(tokens[0]);
	    nsols = parseInt(tokens[1]);
	} else if (tokens.length == n && d.length < n) {
	    d.push(tokens.map(x => parseInt(x)));
	} else if (tokens.length == n+1 && tours.length < nsols) {
	    tours.push(tokens.map(x => parseInt(x)));
	}
    });
    var data = new TSPData(n, d);
    var sols = tours.map(x => new TSPSolution(data, x));
    return [data, sols];
}

function benchmarkOne(path, benchmarkname) {
    var [data, solutions] = readData(path);
    var fname = path.split('/').pop();;
    var nimpr = 0;
    var n = 0;
    var totalTime = 0.0;
    var t1=0.0, t2=0.0;
    for(s in solutions){
	var sol = solutions[s];
	t1 = process.hrtime();
	if (benchmarkname == '2-opt') {
	    n = sol.twoOpt();
	} else if (benchmarkname == 'Or-opt') {
	    n = sol.orOpt();
	} else {
	    console.log('Unknown benchmark: ', benchmarkname);
	}
	t2 = process.hrtime(t1);
	totalTime += t2[1] / 1e9;
	nimpr += n;
    }
    console.log(['javascript', benchmarkname, fname,
		 data.n, solutions.length, nimpr, totalTime].join());
}

function main(args) {
    if (args.length < 3) {
	console.log('USAGE:', args[1], 'tsp_instance_file');
    } else {
	benchmarkname = '2-opt';
	if (args.length > 3) benchmarkname = args[3];
	benchmarkOne(args[2], benchmarkname);
    }
}

main(process.argv);
